Add DSL describing metrics mapping

surik opened the pull request:
Can be used with subscriptions module in `exometer_influxdb` and `exometer_fetch`.
This commit brings `Metricman.Subscription` with two macros: `scope` and `map`. 
See example of using:

```
defmodule MetrcisMapping do
  use Metricman.Subscription

  scope [:ipPools], expose: [:influx] do
    map [:used],  [:ippools, :ip, :total, :used]
    map [:total], [:ippools, :ip, :total, :total]
    scope ["$country"] do
      map [:used],  [:ippools, :ip, "$country", :used]
      map [:total], [:ippools, :ip, "$country", :total]
      map [:req], [:ippools, :ip, "$country", :req], datapoints: [:max, :min]
    end
  end
end
```

will transform to (approximately)

```
defmodule MetrcisMapping do
  def get([:ippools, :ip, :total, :used], type),
    do: {:ok, {[:ipPools, :used], datapoints,  expose: [:influx]}}
  def get([:ippools, :ip, :total, :total], type),
    do: {:ok, {[:ipPools, :total], datapoints,  expose: [:influx]}}
  def get([:ippools, :ip, country, :used], type),
    do: {:ok, {[:ipPools, country, :used], datapoints,  expose: [:influx]}}
  def get([:ippools, :ip, country, :total], type),
    do: {:ok, {[:ipPools, country, :total], datapoints,  expose: [:influx]}}
  def get([:ippools, :ip, country, :req], type),
    do: {:ok, {[:ipPools, country, :total], [:max, :min],  expose: [:influx], datapoints: [:max, :min]}}
  def get(_ , _),
    do: {:error, :not_found}

  ...
end
```

where `datapoints` depends on metric's type or overrided by opts[:datapoints].

surik committed:
Add DSL describing metrics mapping

Can be used with subscriptions module in `exometer_influxdb` and
`exometer_fetch`.
This commit brings `Metricman.Subscription` with two macros: `scope` and
`map`. See example of using:

    defmodule MetrcisMapping do
      use Metricman.Subscription

      scope [:ipPools] do
        map [:used],  [:ippools, :ip, :total, :used]
        map [:total], [:ippools, :ip, :total, :total]
        scope ["$country"] do
          map [:used],  [:ippools, :ip, "$country", :used]
          map [:total], [:ippools, :ip, "$country", :total]
        end
      end
    end

will transform to (approximately)

    defmodule MetrcisMapping do
      def get([:ippools, :ip, :total, :used], type),
        do: {:ok, {[:ipPools, :used], datapoints}}
      def get([:ippools, :ip, :total, :total], type),
        do: {:ok, {[:ipPools, :total], datapoints}}
      def get([:ippools, :ip, country, :used], type),
        do: {:ok, {[:ipPools, country, :used], datapoints}}
      def get([:ippools, :ip, country, :total], type),
        do: {:ok, {[:ipPools, country, :total], datapoints}}
      def get(_ , _),
        do: {:error, :not_found}

      ...
    end

where `datapoints` depends on metric's type.

GalaxyGorilla commented:
For a better interplay with reporter and autosubscriptions I want to to propose another 'Meta' field in this DSL. This field can be an arbitrary Erlang term which can be used to e.g. distinguish subscriptions to different reporters. Example:

```
scope [:metric], [:influx, :prometheus] do
    map [:one],   [:exo, :one]
    map [:two],   [:exo, :two],   [:influx]
    map [:three], [:exo, :three], [:something_else]
```

Here a list of of different TSDBs is used in the 'scope' part to say "only subscribe to those reporters". Within each 'map' this can be overwritten. This translates as follows to code:

```
def get([:exo, :one], type),
    do: {:ok, {[:metric, :one], datapoints, [:influx, :prometheus]}}
  def get([:exo, :two], type),
    do: {:ok, {[:metric, :two], datapoints, [:influx]}}
  def get([:exo, :three], type),
    do: {:ok, {[:metric, :three], datapoints, [:something_else]}}
```

Also a small mapping for datapoints might be a good idea. Not sure how to put this into this DSL nicely, but it would be cool to say "only take those datapoints for this metric" and by default every datapoint possible. This would mean for function types to automatically provide the custom datapoints which is very handy.

surik commented:
In my opinion better to use options list for TSDBs, datapoints and etc... 
Example:

``` elixir
scope [:metric], engine: [:influx, :prometheus] do
    map [:one],   [:exo, :one]
    map [:two],   [:exo, :two],  engine: [:influx], datapoints: [:min, :max]
    map [:three], [:exo, :three], engine: [:something_else]
end
```

And use list in generated functions too:

``` elixir
def get([:exo, :one], type),
    do: {:ok, [:metric, :one], datapoint: [:value], engine: [:influx, :prometheus]}
  def get([:exo, :two], type),
    do: {:ok, [:metric, :two], datapoint: [:min, :max], engine: [:influx]}
  def get([:exo, :three], type),
    do: {:ok, [:metric, :three], datapoint: [:value], engine: [:something_else]}
```

GalaxyGorilla commented:
Looks good! My only concern is the naming ... Maybe something like 'expose' instead of 'engine'? And instead of 'datapoint' -> 'datapoints' ?

surik committed:
Add options for subscription

Example:

    scope [:ipPools], expose: influx do
        map [:used],  [:ippools, :ip, :total, :used]
    end

generates

    get([:ippools. :ip, :total, :used], metric_type, []) ->
      {[:ipPools, :used], datapoints, expose: influx}

Also options can be used for map.

surik commented:
@GalaxyGorilla updated. check it please
